src/lib.rs

```diff
@@ -0,0 +1,452 @@
+// notice and either this complete permission notice or at a minimum a reference
+// to the UPL must be included in all copies or substantial portions of the
+// Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+extern crate elf;
+extern crate byteorder;
+
+mod errors;
+#[macro_use]
+mod util;
+
+use std::path::Path;
+use std::io::Cursor;
+use byteorder::{NativeEndian, ReadBytesExt};
+use errors::{SMParserError, SMParserResult};
+use util::{cursor_skip, cursor_align8, cursor_from_elf};
+
+// We only support this version of the stackmap header for now.
+const STACKMAP_VERSION: u8 = 3;
+
+// Sizes in bytes.
+const SIZE_STACK_SIZE_ENTRY: u8 = 24;
+const SIZE_CONSTANT_ENTRY: u8 = 8;
+const SIZE_LOC_ENTRY: u8 = 12;
+const SIZE_LIVEOUT_ENTRY: u8 = 4;
+
+/// Offsets into the stackmap section.
+const OFFS_STACK_SIZE_ENTRIES: u64 = 16;
+
+/// Represents a single stackmap record entry.
+#[derive(Debug, Eq, PartialEq)]
+pub struct SMRec {
+    id: u64,            // Stackmap ID.
+    offset: u32,        // Stackmap offset from start of containing func.
+}
+
+impl SMRec {
+    /// Get the stackmap ID.
+    pub fn id(&self) -> u64 {
+        self.id
+    }
+
+    /// Get the offset of the stackmap from the start of the containing function.
+    pub fn offset(&self) -> u32 {
+        self.offset
+    }
+}
+
+/// Represents a single function entry.
+#[derive(Debug, Eq, PartialEq)]
+pub struct SMFunc {
+    addr: u64,          // Function address.
+    stack_size: u64,    // Function's stack size.
+}
+
+impl SMFunc {
+    /// Get the function address.
+    pub fn addr(&self) -> u64 {
+        self.addr
+    }
+
+    /// Get the size of the stack of the function.
+    pub fn stack_size(&self) -> u64 {
+        self.stack_size
+    }
+}
+
+/// An iterator over stackmap record entries.
+pub struct SMRecIterator<'a> {
+    elf_file: &'a elf::File,
+    cursor: Option<Cursor<&'a Vec<u8>>>, // Lazily created so that creating the iterator doesn't
+                                         // need to return a `Result`.
+    start_pos: u64,                      // Start position of the cursor.
+    num_stackmaps: u32,
+}
+
+impl<'a> Iterator for SMRecIterator<'a> {
+    type Item = SMParserResult<SMRec>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.cursor.is_none() {
+            self.cursor = Some(itry!(cursor_from_elf(self.elf_file, self.start_pos)));
+        }
+        let mut cursor = self.cursor.as_mut().unwrap();
+
+        if self.num_stackmaps == 0 {
```

Comment:
Shouldn't this go at the start of the function? It doesn't seem to rely on anything earlier in the function?

Reply:
Good idea.